serverless-map-reduce/README.md

```diff
@@ -0,0 +1,74 @@
+```
+
+This should print an approximation of Pi. Feel free to play around with the number of points to achieve better accuracy. Make sure that the number of reduce points is the sum of all the input points.
+
+You can track the progress of the MapReduce job using:
+
+```bash
+$ bn logs public_mr_controller
+```
+
+## A naive serverless MapReduce framework
+
+Our framework has just one function `public_mr_controller` which takes the naive approach of invoking all mapper functions in parallel and waiting for them to complete. Once they complete, it invokes the reducer function and returns the result to the caller.
+
+This approach allows the user to have their map (`public_compute_pi_mapper`) and reduce (`public_compute_pi_reducer`) functions focus purely on computational logic. However, it does not account for failures, retries or laggards. In other examples we use streaming to implement a more robust asynchronous framework.
+
+## Running the asynchronous example
+
+Once you have the naive example up and running, you can proceed with the following steps:
+
+```bash
+$ npm install ioredis
+$ bn deploy public_amr_controller
+$ bn deploy public_amr_mapper
+$ bn trigger setup amr --source-uri "redis+tls://:$REDIS_PASSWORD@$REDIS_HOST:$REDIS_PORT#inputs" --target public_amr_mapper
+```
+
+You can run the example with the async example `job.json` file:
+
+```bash
+$ bn invoke public_amr_controller --json job.json
+```
+
+This will kick off an asynchronous computation, and so will not output and result. You can track the progress of the MapReduce job and see the final estimation of Pi by running:
+
+```bash
+$ bn logs public_amr_mapper
+```
+
+## An asynchronous serverless MapReduce framework
+
+Our async framework uses two functions: `public_amr_controller` and `public_amr_mapper`. We also use a Redis stream to facilitate asynchronous invocation of the mapper function.
+
+The controller function posts invocation requests onto the Redis stream. Each request corresponds to a single mapper job. A Binaris trigger is configured to listen to this Redis stream and invoke the `public_amr_mapper` functions for each data element posted onto the stream.
+
+The mapper function `public_amr_mapper` is a bit more complicated and performs the following steps:
+* Extract the job information from the stream data element.
+* Invoke the user's mapper function specified in the job information. In our case, this is `public_compute_pi_mapper`.
+* Save the output from the mapper onto a Redis HSET.
+* Use Redis atomic increment to increase a count of terminated map jobs.
+* When the count reaches the overall number of map jobs (this will happen only once because Redis increment is atomic), invoke the user's reducer function `public_compute_pi_reducer`.
+* At this point we have the coverall result of the computation and we could store it in Redis or stream it back to the user (using another stream, of course). In this case, we just print it to the log.
```

Comment:
"coverall"

https://www.duluthtrading.com/dw/image/v2/BBNM_PRD/on/demandware.static/-/Sites-dtc-master-catalog/default/dw652a5e1c/images/large/76259_SBL.jpg?sw=600&sh=600&sm=fit

Reply:
"store it in Redis " _> "store it in Redis, "